cranelift/wasm/src/code_translator.rs

```diff
@@ -3579,50 +3581,255 @@ where
-        let iter_ptr = builder.ins().iadd_imm(iter_ptr, i);
-        builder.ins().arm64_st2g(tagged_ptr, iter_ptr);
-        i += 32;
-    }
-    if i < size {
-        let iter_ptr = builder.ins().iadd_imm(iter_ptr, i);
-        builder.ins().arm64_stg(tagged_ptr, iter_ptr);
+    let size_u64: u64 = size as u64;
+    // In Rust, for u64: (a / b) == floor(a / b), which we make use of here
+    let num_required_st2g_loops: u64 = size_u64 / LOOP_UNROLL_BYTES_THRESHHOLD_U64;
+    let num_required_last_stgs: u64 = (size_u64 % LOOP_UNROLL_BYTES_THRESHHOLD_U64) / 16;
+
+    // TODO: cast may be wrong, but i64 is required for use as a Imm64
+    let num_required_st2g_loops_i64: i64 = num_required_st2g_loops as i64;
+
+    if num_required_st2g_loops <= 1 {
+        // WITHOUT_LOOP_UNROLLING_THRESHHOLD
+
+        tag_memory_region_static_without_loop_unrolling_threshhold(
+            iter_ptr, 0, tagged_ptr, size, builder,
+        );
+    } else {
+        // WITH_LOOP_UNROLLING_THRESHHOLD
+
+        // === Block definitions
+        // st2g_loop_condition_block(iter_ptr, st2g_loop_counter)
+        let st2g_loop_condition_block =
+            block_with_params(builder, [ValType::I64, ValType::I64], environ)?;
+
+        // TODO: does this really require iter_ptr and counter as parameters, since they are never modified in the condition
+        // st2g_loop_body_block(iter_ptr, st2g_loop_counter)
+        let st2g_loop_body_block =
+            block_with_params(builder, [ValType::I64, ValType::I64], environ)?;
+
+        // TODO: potential optimization: since this isn't required for all generated code, can we conditionally define this block only when needed?
+        // last_stgs_block()
+        let last_stgs_block = block_with_params(builder, [], environ)?;
+
+        // next_block()
+        let next_block = block_with_params(builder, [], environ)?;
+
+        let st2g_loop_counter = builder.ins().iconst(I64, 0);
+        builder
+            .ins()
+            .jump(st2g_loop_condition_block, &[iter_ptr, st2g_loop_counter]);
+
+        // === st2g_loop_condition_block(iter_ptr, st2g_loop_counter)
+        builder.switch_to_block(st2g_loop_condition_block);
+
+        let iter_ptr = builder.block_params(st2g_loop_condition_block)[0];
+        let st2g_loop_counter = builder.block_params(st2g_loop_condition_block)[1];
+
+        let cond = builder.ins().icmp_imm(
+            IntCC::UnsignedLessThan,
+            st2g_loop_counter,
+            num_required_st2g_loops_i64,
+        );
+        canonicalise_brif(
+            builder,
+            cond,
+            st2g_loop_body_block,
+            &[iter_ptr, st2g_loop_counter],
+            if num_required_last_stgs == 0 {
+                next_block
+            } else {
+                last_stgs_block
+            },
+            &[],
+        );
+
+        // === st2g_loop_body_block(iter_ptr, st2g_loop_counter)
+        builder.switch_to_block(st2g_loop_body_block);
+
```

Comment:
Ups, I actually think there is a bug here, I forgot to save the `iter_ptr` and `st2g_loop_counter` (thought that one won't be needed anymore due to an optimization) branch parameters to new local variables.

Reply:
Should be fixed now though.